File: vss-mobile/android/src/main/java/com/vss/simulator/SimulatorSurfaceView.java
package com.vss.simulator;

import android.content.Context;
import android.hardware.HardwareBuffer;
import android.os.Looper;
import android.util.AttributeSet;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import androidx.annotation.NonNull;

/**
 * Surface view for simulator rendering.
 */
public class SimulatorSurfaceView extends SurfaceView implements SurfaceHolder.Callback2 {

    public SimulatorSurfaceView(Context context) {
        super(context);
        initialize();
    }

    public SimulatorSurfaceView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initialize();
    }

    public SimulatorSurfaceView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initialize();
    }

    private void initialize() {
        this.setAlpha(1.0f);
        this.getHolder().addCallback(this);
    }

    @Override
    public void surfaceCreated(@NonNull SurfaceHolder holder) {
        assert Looper.getMainLooper().isCurrentThread() : "Called from non-UI thread";
        SimulatorBridge.create(holder.getSurface(), getResources().getAssets());
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
        assert Looper.getMainLooper().isCurrentThread() : "Called from non-UI thread";
        SimulatorBridge.resize(width, height);
    }

    @Override
    public void surfaceDestroyed(@NonNull SurfaceHolder holder) {
        assert Looper.getMainLooper().isCurrentThread() : "Called from non-UI thread";
        SimulatorBridge.destroy();
    }

    @Override
    public void surfaceRedrawNeeded(SurfaceHolder holder) {
        assert Looper.getMainLooper().isCurrentThread() : "Called from non-UI thread";
        SimulatorBridge.draw();
    }

    public void postHardwareBuffer(
            int width,
            int height,
            int dataSpace,
            int rotationDegrees,
            HardwareBuffer hardwareBuffer
    ) {
        assert Looper.getMainLooper().isCurrentThread() : "Called from non-UI thread";
        SimulatorBridge.postHardwareBuffer(width, height, dataSpace, rotationDegrees, hardwareBuffer);
        SimulatorBridge.draw();
    }

    public void postRgba(int width, int height, java.nio.ByteBuffer pixels) {
        assert Looper.getMainLooper().isCurrentThread() : "Called from non-UI thread";
        SimulatorBridge.postRgba(width, height, pixels);
        SimulatorBridge.draw();
    }

    public void postSettings(String jsonString) {
        assert Looper.getMainLooper().isCurrentThread() : "Called from non-UI thread";
        SimulatorBridge.postSettings(jsonString);
        SimulatorBridge.draw();
    }

    public String querySettings() {
        assert Looper.getMainLooper().isCurrentThread() : "Called from non-UI thread";
        return SimulatorBridge.querySettings();
    }
}

File: vss-android/app/src/main/kotlin/com/vss/simulator/SimulatorSurfaceView.kt
package com.vss.simulator

import android.content.Context
import android.hardware.HardwareBuffer
import android.os.Looper
import android.util.AttributeSet
import android.view.SurfaceHolder
import android.view.SurfaceView
import java.nio.ByteBuffer

/** Surface view for simulator rendering. */
class SimulatorSurfaceView : SurfaceView, SurfaceHolder.Callback2 {
    constructor(context: Context) : super(context) {
        initialize()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        initialize()
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
        super(context, attrs, defStyleAttr) {
        initialize()
    }

    private fun initialize() {
        alpha = 1.0f
        holder.addCallback(this)
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        assertMainThread()
        SimulatorBridge.create(holder.surface, resources.assets)
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        assertMainThread()
        SimulatorBridge.resize(width, height)
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        assertMainThread()
        SimulatorBridge.destroy()
    }

    override fun surfaceRedrawNeeded(holder: SurfaceHolder) {
        assertMainThread()
        SimulatorBridge.draw()
    }

    fun postHardwareBuffer(
        width: Int,
        height: Int,
        dataSpace: Int,
        rotationDegrees: Int,
        hardwareBuffer: HardwareBuffer,
    ) {
        assertMainThread()
        SimulatorBridge.postHardwareBuffer(width, height, dataSpace, rotationDegrees, hardwareBuffer)
        SimulatorBridge.draw()
    }

    fun postRgba(width: Int, height: Int, pixels: ByteBuffer) {
        assertMainThread()
        SimulatorBridge.postRgba(width, height, pixels)
        SimulatorBridge.draw()
    }

    fun postSettings(jsonString: String) {
        assertMainThread()
        SimulatorBridge.postSettings(jsonString)
        SimulatorBridge.draw()
    }

    fun querySettings(): String {
        assertMainThread()
        return SimulatorBridge.querySettings()
    }

    private fun assertMainThread() {
        assert(Looper.getMainLooper().isCurrentThread) { "Called from non-UI thread" }
    }
}
